### ISSUE TITLE
[apigateway] Unable to add accurate resource policy

### ISSUE THREAD

**Issue by brettswift** [aws/aws-cdk#2387]
The resource policy is readonly in the RestApi.  It's only injectable at the constructor.  This is a problem because I can't get the ARN into the policy to control method level invocations.  It seems like there's a chicken/egg situation. 

Also - changing the policy doesn't seem to kick off a deployment of the API, meaning the changes have no effect.  Adding the policy contents to the hash of the deployment logical id seems logical. (This might belong in it's own issue).

What I'm really trying to accomplish is a customization of the RestApi (like LambdaRestApi did) but for a 'flavoured' . regional apigateway with some security features that I can share amongst the team. 

Is making this mutable going to interfere with the template prepare() activities?  

Opening this up will also let us apply policy statements after resources / methods have been added to the api. 

```
      new apigateway.RestApi(this, 'testApi', {
        policy: new iam.PolicyDocument()
                .addStatement(new iam.PolicyStatement(iam.PolicyStatementEffect.Deny)
                    .addAction("execute-api:Invoke")
                    .addAnyPrincipal()
                    .addResource(`arn:aws:execute-api:${cdk.Aws.region}:*:<how-to-get-id>/prod/`)
                )
      })

```

**Comment by orangewise** [comment]
I think apigateway will only update if the `Logical ID` of `AWS::ApiGateway::Deployment` is changed...

**Comment by brettswift** [comment]
@orangewise yes, but my point is that other constructs will trigger a new deployment when they change, like adding a new method for example.   I would expect a change to the policy to trigger a deployment as well. 

I'm doing this manually, but I think the policy itself can take care of this.

**Comment by hoegertn** [comment]
@eladb 	anything on this topic? The not-refreshed deployment is not a gap. Instead, it should qualify as a bug.

**Comment by NGL321** [comment]
Hey @brettswift,

Sorry for so long without a response, I have redesignated this as a bug (thanks @hoegertn), and someone should look into it soon.

Just for confirmation, are you still encountering this issue?

**Comment by brettswift** [comment]
@NGL321  I haven't been working with api gateway much lately, so I can't confirm.  If this isn't reproducible let me know and I'll try to get a sample repo up.

**Comment by botlin** [comment]
@brettswift I run into the same issue and i was able to solve it.
Just specify your resource like the template given in the AWS Console.

`execute-api:/{{stageNameOrWildcard}}/{{httpVerbOrWildcard}}/{{resourcePathOrWildcard}}`

In CDK Like this
`.addResources("execute-api:/*/POST/myAPI/myEndpoint")`

AWS will add your account, region and stack id automatically.
You can see the same behavior in the console when you create a new resource policy in the API
manually.

**Comment by chillitom** [comment]
```
        const api = new apigateway.RestApi(this, 'api', {
            restApiName: 'API',
            policy: new iam.PolicyDocument({
                statements: [
                    new iam.PolicyStatement({ actions: ['execute-api:*'], resources: ['execute-api:/*'], effect: iam.Effect.DENY }),
                    new iam.PolicyStatement({
                        actions: ['execute-api:*'],
                        resources: ['execute-api:/*'],
                        effect: iam.Effect.ALLOW,
                        principals: [
                            authenticatedRole
                        ],
                    }),
                ]
            })
        })
```

Adding the above policy causes the deployment to hang on deploying the API and it has to be cancelled in the cloudformation console.  Any ideas what's up with this declaration?

**Comment by prasadgunasekaran** [comment]
i use python cdk and it hangs for me too, any updates on this particular case.

**Comment by marcogrcr** [comment]
This works just fine for me with `npm`'s `"@aws-cdk/aws-apigateway": "1.23.0"`:

```ts
import { RestApi } from "@aws-cdk/aws-apigateway";
import { AnyPrincipal } from "@aws-cdk/aws-iam";
import { Token } from "@aws-cdk/core";

const apiPolicy = new PolicyDocument({
  statements: [
    new PolicyStatement({
      actions: ["execute-api:Invoke"],
      // prod is `RestApi`'s default stage name, `*` means ANY method, `/public/*` means any path that starts with `/public`
      resources: ["execute-api:/prod/*/public/*"],
      principals: [new AnyPrincipal()]
    })
  ]
});

const api = new RestApi(this, "Api", {
  // ...
  policy: apiPolicy
});

// ensure a new API deployment occurs when the API policy changes
api.latestDeployment?.addToLogicalId(Token.asAny(apiPolicy));
```

**Comment by nija-at** [comment]
There are two parts to the original issue.

1. Changes in policy does not trigger a new deployment. This was [fixed](https://github.com/aws/aws-cdk/commit/1b3aeaf27b6491bc6d9c7c83dc7b03b8cd488534#diff-f555223d437a822094c168c0b2aee75c) a few months ago and released in `1.35.0`.

2. Ability to add statements to the RestApi policy after initialization. I've opened a separate feature request to track this - https://github.com/aws/aws-cdk/issues/8781.

Marking this as closing soon. Let me know if there's something else I've missed here.
